test/common/stats/stat_test_utility.h

```diff
@@ -27,6 +23,79 @@ bool hasDeterministicMallocStats();
  */
 void forEachSampleStat(int num_clusters, std::function<void(absl::string_view)> fn);
 
+// Tracks memory consumption over a span of time. Test classes instantiate a
+// MemoryTest object to start measuring heap memory, and call consumedBytes() to
+// determine how many bytes have been consumed since the class was instantiated.
+//
+// That value should then be passed to EXPECT_MEMORY_EQ and EXPECT_MEMORY_LE,
+// defined below, as the interpretation of this value can differ based on
+// platform and compilation mode.
+class MemoryTest {
+public:
+  // There are 3 cases:
+  //   1. Memory usage API is available, and is built using with a canonical
+  //      toolchain, enabling exact comparisons against an expected number of
+  //      bytes consumed. The canonical environment is Envoy CI release builds.
+  //   2. Memory usage API is available, but the current build may subtly differ
+  //      in memory consumption from #1. We'd still like to track memory usage
+  //      but it needs to be approximate.
+  //   3. Memory usage API is not available. In this case, the code is executed
+  //      but no testing occurs.
+  enum class Mode {
+    Disabled,    // No memory usage data available on platform.
+    Canonical,   // Memory usage is available, and current platform is canonical.
+    Approximate, // Memory usage is available, but variances form canonical expected.
+  };
+
+  MemoryTest() : memory_at_construction_(Memory::Stats::totalCurrentlyAllocated()) {}
+
+  /**
+   * @return the memory execution testability mode for the current compiler, architecture,
+   *         and compile flags.
+   */
+  static Mode mode();
+
+  size_t consumedBytes() const {
+    return Memory::Stats::totalCurrentlyAllocated() - memory_at_construction_;
+  }
+
+private:
+  const size_t memory_at_construction_;
+};
+
+// Compares the memory consumed against an exact expected value, but only on
+// canonical platforms, or when the expected value is zero. Canonical platforms
+// currently include only for 'release' tests in ci. On other platforms an info
+// log is emitted, indicating that the test is being skipped.
+#define EXPECT_MEMORY_EQ(consumed_bytes, expected_value)                                           \
```

Comment:
Does this two macro have to be macro, instead of a static method?

Reply:
Yes, if you want to have accurate _test.cc line numbers when tests fail, and I personally really appreciate that :)